plugins/woocommerce/includes/wc-product-functions.php

```diff
@@ -282,6 +282,47 @@ function wc_product_post_type_link( $permalink, $post ) {
 }
 add_filter( 'post_type_link', 'wc_product_post_type_link', 10, 2 );
 
+/**
+ * Ensure that the product_cat value determined in `wc_product_post_type_link` is the canonical value.
+ *
+ * If other values are used in this part of the permalink, it will be redirected.
+ *
+ * @return void
+ */
+function wc_product_canonical_redirect(): void {
+	global $wp_rewrite;
+
+	if (
+		! did_action( 'woocommerce_init' )
```

Comment:
I think this may be superfluous now we're placing the logic inside WooCommerce itself ... then again, perhaps the safest thing is to leave it (from a defensive code perspective, as we often come up against unexpected conditions) 🤔

Reply:
Yeah, I think there are a lot of other places where we make this `did_action` check within WC... I'm inclined to leave it there.